### PR TITLE
MSC3706: Extensions to `/_matrix/federation/v2/send_join/{roomId}/{eventId}` for partial state

### PR DESCRIPTION
**Pull request by richvdh**
Rendered: https://github.com/matrix-org/matrix-doc/blob/rav/proposal/partial_state_on_join/proposals/3706-partial-state-in-send-join.md

Server-side synapse implementation is in https://github.com/matrix-org/synapse/pull/11967.
Client-side parsing is https://github.com/matrix-org/synapse/pull/12011; the real implementation is an 8-month epic, but https://github.com/matrix-org/synapse/blob/v1.68.0/synapse/handlers/federation.py#L616-L625 is a reasonable place to start looking.

[FCP comment](https://github.com/matrix-org/matrix-spec-proposals/pull/3706#issuecomment-1274758677)

### CODE REVIEW

**Comment by reivilibre** [comment]
Seems fine, though for completeness I wanted to ask: is it worth considering allowing other kinds of state events to be omitted in the future?
e.g. Should the responding homeserver declare the set of state events that have been left out?
I'm not super concerned about doing that *right now*, of course, but would be interesting if you had any thoughts about how to 'leave the doors open' for that extension, as it were.
Membership events are presumably the bulk *right now*, but I can somewhat imagine that some other applications (mainly of the 'beyond chat' variety?) might have similar bulky state events.

**Comment by richvdh** [comment]
hrm, interesting idea. Will bear it in mind as we see how this unfolds.

**Comment by squahtx** [comment]
I think it could be made clearer that `partial_state: true` in the response implies `servers_in_room` is present, otherwise `servers_in_room` may or may not be present.

File: proposals/3706-partial-state-in-send-join.md [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+   > Any events returned within `state` can be omitted from `auth_chain`.
+
+   For example: the `m.room.create` event is part of the room state, so
+   must be included in `state`. However, it also forms part of the auth chain
+   for all of the returned events, so in the current spec, must *also* be
+   included in `auth_chain`. However, this is redundant, so we should omit it
+   for calling servers which opt into that via the `partial_state` query param.
+
+ * `servers_in_room`: A new field of type `[string]`, listing the servers
```
Comment (richvdh) [inline proposals/3706-partial-state-in-send-join.md:70]: @squahtx says:

> I think it could be made clearer that partial_state: true in the response implies servers_in_room is present, otherwise servers_in_room may or may not be present.

(please remember to use comments on the diff, so that we get some basic threading)

File: proposals/3706-partial-state-in-send-join.md [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+ * `partial_state`: a new boolean field is added. This should be set to `true`
+   to indicate that partial state is being returned. It must otherwise be set
+   to `false` or omitted.
+
+ * `state`: if partial state is being returned, then a subset of the full room
+   state, rather than the complete room state. In particular, the following
+   state should be returned:
+
+     * any state with event type other than `m.room.member`.
+     * TODO: anything else?
```
Comment (richvdh) [inline proposals/3706-partial-state-in-send-join.md:48]: @reivilibre says:

> Seems fine, though for completeness I wanted to ask: is it worth considering allowing other kinds of state events to be omitted in the future?
e.g. Should the responding homeserver declare the set of state events that have been left out?
I'm not super concerned about doing that right now, of course, but would be interesting if you had any thoughts about how to 'leave the doors open' for that extension, as it were.
Membership events are presumably the bulk right now, but I can somewhat imagine that some other applications (mainly of the 'beyond chat' variety?) might have similar bulky state events.

Comment (richvdh) [reply proposals/3706-partial-state-in-send-join.md:48]: hrm, interesting idea. Will bear it in mind as we see how this unfolds.

File: proposals/3706-partial-state-in-send-join.md (hunk 1 above)
Comment (richvdh) [reply proposals/3706-partial-state-in-send-join.md:70]: fair. I've added a note to this effect in [195ba79](https://github.com/matrix-org/matrix-doc/pull/3706/commits/195ba7918a8d06d49029544cc6519b5e73aa18d3).

File: proposals/3706-partial-state-in-send-join.md (hunk 2 above)
Comment (richvdh) [reply proposals/3706-partial-state-in-send-join.md:48]: we've somewhat baked in the assumption that we're only omitting membership events for now. We can revisit this in future MSCs.

**Comment by richvdh** [comment]
I believe this is ready for review.

**Comment by richvdh** [comment]
@mscbot fcp merge

File: proposals/3706-partial-state-in-send-join.md [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+## Proposal
+
+[`PUT /_matrix/federation/v2/send_join/{roomId}/{eventId}/send_join`](https://spec.matrix.org/v1.2/server-server-api/#put_matrixfederationv2send_joinroomideventid)
+is extended to support "partial state" in its responses. This involves the
+following changes.
+
+### New query parameter
+
+`partial_state` is added as a new query parameter. This can take the values
+`true` or `false`; other values should be rejected with an HTTP 400 error.
```
Comment (turt2live) [inline proposals/3706-partial-state-in-send-join.md:26]: Presumably also with an appropriate matrix-standard error code/object.

File: proposals/3706-partial-state-in-send-join.md [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+The following changes are made to the response:
+
+ * `partial_state`: a new boolean field is added. This should be set to `true`
+   to indicate that partial state is being returned. It must otherwise be set
+   to `false` or omitted.
+
+ * `state`: if partial state is being returned, then state events with event
+   type `m.room.member` are omitted from the response. (All other room state is
+   returned as normal.)
```
Comment (turt2live) [inline proposals/3706-partial-state-in-send-join.md:45]: Why specifically member events? With MSC3401 and friends we're about to have a ton of other membership-like state events which should be excluded for performance reasons, or because the relevant call is concluded.

Edit during review: The MSC makes mention of altering the list of omitted state events as a deferred action, however it's not spelled out why membership events are neccessarily the problem here. The intro makes reference to events as a whole, but not membership.

File: proposals/3706-partial-state-in-send-join.md [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+None at present.
+
+## Alternatives
+
+ * In future, the list of event types to omit could be expanded. (Some rooms
+   may have large numbers of other state events). For now, this has been
+   descoped, but could be revisited in future MSCs.
+ 
+## Security considerations
```
Comment (turt2live) [inline proposals/3706-partial-state-in-send-join.md:86]: There's a potential concern I'll discuss internally first then copy here afterwards.

File: proposals/3706-partial-state-in-send-join.md (hunk 3 above)
Comment (richvdh) [reply proposals/3706-partial-state-in-send-join.md:26]: yes.

File: proposals/3706-partial-state-in-send-join.md (hunk 4 above)
Comment (richvdh) [reply proposals/3706-partial-state-in-send-join.md:45]: In short because, of the 200K events returned by a `send_join` to Matrix HQ, 199950 of them are membership events.

I'll update the text to say this in due course.

File: proposals/3706-partial-state-in-send-join.md [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+free to support it for some rooms and not others.
+
+The following changes are made to the response:
+
+ * `partial_state`: a new boolean field is added. This should be set to `true`
+   to indicate that partial state is being returned. It must otherwise be set
+   to `false` or omitted.
+
+ * `state`: if partial state is being returned, then state events with event
+   type `m.room.member` are omitted from the response. (All other room state is
```
Comment (benkuly) [inline proposals/3706-partial-state-in-send-join.md:44]: Would it make sense to make this configurable? A parameter `required_state_events` could indicate, which state events are needed by the server. On the one hand, this would make the API more explicit, which events are in the response and on the other hand this would allow rooms, which have many state events of another type than `m.room.member`. The latter, for example, could be rooms containing a shared document within the state.

Comment (benkuly) [reply proposals/3706-partial-state-in-send-join.md:44]: I know, that `Alternatives` says, that this should be another MSC, but I think we should define an API as clean as possible. This MSC ssems to be a fix for one special use case and not for the problem behind it.

File: proposals/3706-partial-state-in-send-join.md (hunk 5 above)
Comment (richvdh) [reply proposals/3706-partial-state-in-send-join.md:86]: For visibility, the concern was around what happens if the remote server withholds some state in the room.

This is an existing problem: you are putting some trust in the resident server when you join a room. This MSC isn't changing that, so I don't think it's relevant.

File: proposals/3706-partial-state-in-send-join.md (hunk 6 above)
Comment (richvdh) [reply proposals/3706-partial-state-in-send-join.md:44]: the counterpoint is that it's better to land a simple change, rather than wait until you've made it as general as possible, which could take a long time.

In your example: how does a server know which events should be in `required_state_events`?

File: proposals/3706-partial-state-in-send-join.md (hunk 3 above)
Comment (richvdh) [reply proposals/3706-partial-state-in-send-join.md]: updated.

File: proposals/3706-partial-state-in-send-join.md (hunk 4 above)
Comment (richvdh) [reply proposals/3706-partial-state-in-send-join.md]: updated.

File: proposals/3706-partial-state-in-send-join.md (hunk 6 above)
Comment (benkuly) [reply proposals/3706-partial-state-in-send-join.md]: Just to understand: Why is it more complex? Is there some special "logic" which makes it a difference if member events or all events are lazy loaded? Shouldn't the effects on der DAG be the same?

In my example: Why not default it to `every_state_event_type_from_the_spec minus m.room.member`? Or define it as `*` and introduce a `lazy_state_events` with `m.room.member`. In a future MSC this could be defined by the client, who wants to join.

Comment (richvdh) [reply proposals/3706-partial-state-in-send-join.md]: > Just to understand: Why is it more complex? Is there some special "logic" which makes it a difference if member events or all events are lazy loaded? Shouldn't the effects on der DAG be the same?

well, obviously the server needs to know which events have been omitted, so along with tracking which rooms have partial state, we now need to track that. It's not hard, it's just extra stuff to do.

> (various suggestions about how to decide which events to omit)

Sure, there are various possibilities. But then we'd have a big discussion about what the exact perfect negotiation system is, then we'd have to write the code, and none of it would bring any advantages right now. Meanwhile, want to get this out in the wild so people can use it.

Comment (benkuly) [reply proposals/3706-partial-state-in-send-join.md]: As far as I can see there are already some beyond-chat applications right now, which do use state events massively and do not scale well on room joins too.

I think that this API will make problems beeing backward compatible, when a more generic API is introduced. Calling it `partial_state` could mean everything. It would be cleaner to call it `partial_member_state` so this could be deprecated in a more generic API.

My comment on "It's not hard, it's just extra stuff to do.":  Do It Right Or Do It Twice :wink: or I just repeat myself:
> This MSC and API sems to be a fix for one special use case and not for the problem behind it.

File: proposals/3706-partial-state-in-send-join.md [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+ * In future, the list of event types to omit could be expanded. (Some rooms
+   may have large numbers of other state events).
+   
+   Currently, `m.room.member` events are by far the biggest problem. For
+   example, a `/send_join` request for Matrix HQ returns approximately 85000
+   events in `state`, of which all but 44 are `m.room.member`. 
+
+   In order to reduce the scope of the change, we have therefore decided to
+   focus on `m.room.member` events for now. Future MSCs might provde a
+   mechanism for omitting other event types.
```
Comment (dbkr) [inline proposals/3706-partial-state-in-send-join.md:93]: Some rough idea of how we'd get here in the future could be useful: if we later omitted some other event (MSC3401 call member events could end up being quite numerous, for example) how would the joining server know what events it had to catch up on later? Basically I'm just wondering if we'd be better off making it `omit=m.room.member` or just `omit_member_events=true`.

Comment (richvdh) [reply proposals/3706-partial-state-in-send-join.md:93]: this seems to be about the same discussion as [above](https://github.com/matrix-org/matrix-spec-proposals/pull/3706#discussion_r998573694).

I think there are a number of different ways to go. The requesting server could provide a list of state it cares about, or doesn't care about. The responding server could provide a list of state it omitted, or a list of state it included. I was pretty much trying to avoid making a decision on this for now, but perhaps you're right, it needs refining.

It'll need someone else from the Synapse team to pick it up though. @erikjohnston do you have any thoughts on this?

Comment (erikjohnston) [reply proposals/3706-partial-state-in-send-join.md:93]: TBH, I'm disinclined to make further changes unless they really make a difference. Personally, I think if we add a parameter later that allows customising the behaviour it's perfectly reasonable to default to just omitting membership events, i.e. I don't think ignoring this question for now really binds our hands in any way?

Although, I guess the question is how does the requesting server know what the other server has omitted? If it sends a new parameter it needs to know if the other server understood it or not, but that feels like something we can just add in the response?

Comment (benkuly) [reply proposals/3706-partial-state-in-send-join.md:93]: I want to understand what the problem on the synapse-side is? What exactly is the difference in implementing a generic apporach? Should't it be the same as omitting just one event type? I just don't understand the problem here. It feels more like a "we do not have time and desire to change this". And if this is the case, then why do we have MSCs at all, when this is an accepted argument :thinking:

File: proposals/3706-partial-state-in-send-join.md (hunk 6 above)
Comment (richvdh) [reply proposals/3706-partial-state-in-send-join.md]: closing in favour of the other thread where we've started this discussion again.

File: proposals/3706-partial-state-in-send-join.md (hunk 7 above)
Comment (richvdh) [reply proposals/3706-partial-state-in-send-join.md:93]: >  Should't it be the same as omitting just one event type?

Er, no? We need to keep track of which event types are omitted. As I said before.

> It feels more like a "we do not have time and desire to change this".

Sure, that's a large factor, yes. Changing this will require work, which means we can't spend time doing other development.

> And if this is the case, then why do we have MSCs at all, when this is an accepted argument

You seem to be saying we shouldn't bother with MSCs at all unless make them fix all the world's problems at once? This seems like naive thinking. In short: "Perfect is the enemy of good".

The MSC process is about discussion. We're having a discussion about whether we need to make this more general. So far, I've not been convinced by the arguments that we should, but the decision has not yet been made, and other members of the SCT may yet weigh in on the subject. Voices from the community will be considered.

If you (a) feel strongly that this should be changed, and (b) that it is trivial to do so, one way to help the argument would be to demonstrate it: I'm sure the Synapse team would welcome a PR to make the change.

Comment (richvdh) [reply proposals/3706-partial-state-in-send-join.md:93]: another good reason not to extend this beyond `m.room.member` for now:

Currently, the Client-Server api supports lazy-loading `m.room.members` events (https://spec.matrix.org/v1.4/client-server-api/#lazy-loading-room-members). If we know that we have all the state except `m.room.member`, then we can continue to  serve `/sync` and other requests provided the client enables `lazy_load_members`.

In other words: extending this proposal beyond MSC3706 would require more substantial changes to the client-server API.

Comment (benkuly) [reply proposals/3706-partial-state-in-send-join.md:93]: I understand your arguments, although I still think it's a mistake to specify it badly first rather than right away. But it's okay for me, when the parameter get's an explicit name, what it does, like the mentioned `omit_member_events`.

Comment (richvdh) [reply proposals/3706-partial-state-in-send-join.md:93]: yeah we can certainly give it a better name.

**Review by anoadragon453 — APPROVED** [review]

File: proposals/3706-partial-state-in-send-join.md [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+free to support it for some rooms and not others.
+
+The following changes are made to the response:
+
+ * `partial_state`: a new boolean field is added. This should be set to `true`
+   to indicate that partial state is being returned. It must otherwise be set
+   to `false` or omitted.
+
+ * `state`: if partial state is being returned, then state events with event
+   type `m.room.member` are omitted from the response. All other room state is
```
Comment (uhoreg) [inline proposals/3706-partial-state-in-send-join.md:45]: Are all `m.room.member` events required to be omitted, or is the server allowed to include some of them if it feels like it?  If they're required to be omitted, what should the requesting server do if it finds some `m.room.member` events?

File: proposals/3706-partial-state-in-send-join.md [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+   must be included in `state`. However, it also forms part of the auth chain
+   for all of the returned events, so in the current spec, must *also* be
+   included in `auth_chain`. However, this is redundant, so we should omit it
+   for calling servers which opt into that via the `partial_state` query param.
+
+ * `servers_in_room`: A new field of type `[string]`, listing the servers
+   active in the room (ie, those with joined members) before the join.
+
+   This is to be used by the joining server to send outgoing federation
+   transactions while it synchronises the full state.
```
[the remark this answers is not in the document]
Comment (richvdh) [reply proposals/3706-partial-state-in-send-join.md:73]: Well, that's the broader topic of msc3902. I'm trying to keep this MSC laser-focussed on the changes to the API, rather than the gnarlier details of how servers are meant to use it.

File: proposals/3706-partial-state-in-send-join.md (hunk 8 above)
Comment (richvdh) [reply proposals/3706-partial-state-in-send-join.md:45]: Good question. The short answer is that it doesn't currently matter, and a receiving server will auth and persist the events but won't take a whole lot of notice of them: any operations that require it to have `m.room.member` events will block until the resync.

Do you think this needs specifying more tightly?

File: proposals/3706-partial-state-in-send-join.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (uhoreg) [reply proposals/3706-partial-state-in-send-join.md:73]: Right, maybe something like 

```suggestion
   transactions while it synchronises the full state, as outlined in [MSC3902](https://github.com/matrix-org/matrix-spec-proposals/pull/3902).
```

would be helpful for people like me who missed the pointer to MSC3902 in the introduction.

File: proposals/3706-partial-state-in-send-join.md [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+The following changes are made to the response:
+
+ * `partial_state`: a new boolean field is added. This should be set to `true`
+   to indicate that partial state is being returned. It must otherwise be set
+   to `false` or omitted.
+
+ * `state`: if partial state is being returned, then state events with event
+   type `m.room.member` are omitted from the response. All other room state is
+   returned as normal. (See 'Alternatives' for discussion on why only
+   `m.room.member` events are omitted.)
```
Comment (DMRobertson) [inline proposals/3706-partial-state-in-send-join.md:47]: Per https://github.com/matrix-org/synapse/issues/12995#issuecomment-1290607134: we want to include heroes' memberships here, and include their memberships' auth chain too.

Comment (DMRobertson) [reply proposals/3706-partial-state-in-send-join.md:47]: Rich suggests speccing this as a separate change in the future.

Comment (richvdh) [reply proposals/3706-partial-state-in-send-join.md:47]: My idea was to land this MSC as-is, asap (since matrix.org is already implementing it as-is), and do things like heroes in a future MSC, but I'm not really able to champion this either way right now.

File: proposals/3706-partial-state-in-send-join.md (hunk 7 above)
Comment (dbkr) [reply proposals/3706-partial-state-in-send-join.md:93]: @mscbot concern Decide whether to rename the flag or not

File: proposals/3706-partial-state-in-send-join.md (hunk 10 above)
Comment (DMRobertson) [reply proposals/3706-partial-state-in-send-join.md:47]: #3943.

File: proposals/3706-partial-state-in-send-join.md (hunk 7 above)
Comment (richvdh) [reply proposals/3706-partial-state-in-send-join.md:101]: actually in 12e11ec1 I have switched to `omit_members` etc, for consistency with `lazy_load_members` in the C-S API.

Comment (dbkr) [reply proposals/3706-partial-state-in-send-join.md:101]: @mscbot resolve Decide whether to rename the flag or not

**Comment by richvdh** [comment]
Spec PR: https://github.com/matrix-org/matrix-spec/pull/1393
